src/torrent.cpp

```diff
@@ -1138,6 +1138,44 @@ bool is_downloading_state(int const st)
 		if (m_peer_list) m_peer_list->clear();
 	}
 
+	void torrent::set_sequential_range(piece_index_t first_piece, piece_index_t last_piece)
+	{
+		if (!has_picker()) {
+			if (!valid_metadata() || !m_connections_initialized) return;
+			else if (!m_have_all
+				|| settings().get_int(settings_pack::suggest_mode)
+				== settings_pack::suggest_read_cache)
+				need_picker();
+			else return;
+		}
+		if (first_piece > m_torrent_file->last_piece() || last_piece > m_torrent_file->last_piece()) return;
+		TORRENT_ASSERT_PRECOND(first_piece <= m_torrent_file->last_piece() && last_piece <= m_torrent_file->last_piece());
+		m_sequential_download = true;
+		m_picker->set_sequential_range(first_piece, last_piece);
+	}
+
+	void torrent::set_sequential_range(piece_index_t first_piece)
+	{
+		if (!has_picker()) {
+			if (!valid_metadata() || !m_connections_initialized) return;
+			else if (!m_have_all
+				|| settings().get_int(settings_pack::suggest_mode)
+				== settings_pack::suggest_read_cache)
+				need_picker();
+			else return;
+		}
+		// m_torrent_file->files() is file_storage
+		// find the last file in piece `first_piece`
+		auto const f = m_torrent_file->files().map_block(first_piece, m_torrent_file->files().piece_size(first_piece) - 1, 1).back().file_index;
+
+		// find the last piece in file `f`
+		std::int64_t piece_size = m_torrent_file->files().piece_length();
+		auto const last_piece = piece_index_t(int((m_torrent_file->files().file_offset(f) + m_torrent_file->files().file_size(f) - 1) / piece_size));
```

Comment:
It's strange but that `piece_index_t()` constructor doesn't automatically cast the expression result without a warning error, even though the `int()` constructor casts it without a warning error.

Is `int()` a constructor or a cast function in C++?